---
title: "Liability distributions with LT-FH"
output: rmarkdown::html_vignette
description: >
  An overview and explanation of the methodology used for drawing liabilities.
vignette: >
  %\VignetteIndexEntry{liability-distribution}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

This vignette explains what assumptions are made during both individual and family simulations, furhtermore it shows how we model the liability distributions based on the work by Hujoel et al (2020). It will go into both the liability threshold model and the liability threshold and family history model (LT-FH).

## Simulation
In the geneference package it is possible to simulate both individual and family genetic data, to be used for further analysis. Each simulation will have $n$ individuals or families with $m$ individual Single-Nucleotide Polymorphisms (SNPs), with $c$ causal SNPs. Only the genotypes of the target will be saved in ..., but the liabilities and case/control status of each individual family member will be saved in .... The method of obtaining these will be explain in the section 'Liability threshold model'. 
For the individual simulation each SNP will represent the number of minor allele with ... being sampled from a binomial distribution $SNP_j \sim binom(2,MAF_j)$, with a success probability of the minor allele frequency ($MAF$), being sampled from a uniform distribution $MAF_j \sim unif(0.01,0.49)$, which will be saved in .... Furthermore, the effect size of each causal SNP is being sampled from a normal distribution $\beta_j \sim N(0,\frac{h^2}{c})$, where $h^2$ is the heritability of trait, meaning the degree of variation in a phenotypic trait in a population, and $c$ is the number of causal SNPs. The non-causal SNPs will have an effect size of $0$. The effect sizes will also be saved in .... For the family simulation the parents genotypes are being sampled the same way as the in the individual simulation, but to get the targets genotypes and also the siblings the average of the parents genotypes are used for each SNP. This is done by ....

## Liability threshold model
The liability threshold model assigns a phenotype, case or control, to each person based on their full liability $\epsilon$, which can be divided into two parts, namely genetic liability ($\epsilon_{g}$) and environment $\epsilon_{e}$. As the SNPs are simulated independently no linkage disequilibrium will be taken into consideration and the effect size is assumed to have a linear relationship to the minor allele, meaning that the effect size is constant.The genetic liability is therefore calculated by multiplying the SNPs by their effect size and summing them $\epsilon_g = X_i \beta$, where $X_i$ is the i'th persons SNPs and $\beta$ is the effect sizes. It is assumed that $\epsilon_g \sim N(0,h^2)$, where $h^2$ is the heritability. The liability from the environment is sampled and is assumed to follow a normal distribution, $\epsilon_e \sim N(0,1-h^2)$, which makes the full liability $\epsilon = \epsilon_g+\epsilon_e \sim N(0,1)$. The liability theshold model assigns a case to a person if $\epsilon\geq T$, with $P(\epsilon\geq T) = k$, where $k$ is the prevalence of trait, e.g $5\%$ and $T$ is the threshold of disease liability, and control otherwise. The liabilities and the case/control status is saved in ....

## LT-FH model
The LT-FH method relies on the liability threshold model, but expands it to be able to take case/control status and/or family history into consideration. This is done by calculating posterior mean genetic liability for each target person, where the case/control status for the person and the family is taken into account. The advantage of doing this is to use prior knowledge about the family history to make a more accurate estimate, and also instead of using a binary variable for later analysis having a wider range of values. This is imortant as there are a difference between having a liability just above the threshold of disease liability ($T$) and having a liability way above, which the liability threshold model doesn't consider.

To model a family and their liabilities, it is assumed that it follows a multivariate normal distribution, whith, for two family members, the $cov(\epsilon_1, \epsilon_2) = K_{12}h^2$, where $K_{12}$ is the coefficient of the relationship, for example $1/2$ for a parent-offspring pair. Therefore, the liabilities of a family follows:
$$
\begin{pmatrix}
\epsilon_{o,e} \\ \epsilon_{o,g} \\ \epsilon_{p1} \\ \epsilon_{p2} \\ \epsilon_{s}
\end{pmatrix} \sim
MVN_5 \begin{pmatrix}
\begin{pmatrix}
0 \\ 0 \\ 0 \\ 0 \\ 0
\end{pmatrix},
\begin{pmatrix}
1-h^2 & 0 & 0 & 0 & 0 \\
0 & h^2 & 0.5h^2 & 0.5h^2 & 0.5h^2 \\
0 & 0.5h^2 & 1 & 0 & 0.5h^2 \\
0 & 0.5h^2 & 0 & 1 & 0.5h^2 \\
0 & 0.5h^2 & 0.5h^2 & 0.5h^2 & 1
\end{pmatrix}
\end{pmatrix}
$$
where $\epsilon_{o,e}$ and $\epsilon_{o,g}$ is the environmental and genetic liability of the target, $\epsilon_{p1}$ and $\epsilon_{p2}$ is the full liability of the parents, and $\epsilon_s$ is the full liability of the sibling(s). This is only for one sibling, but can be expanded to an arbitrary number of siblings. 

To be able to include the case/control status for the family, each person is assigned a configuration. The configuration is created based on the case/control status of the target, the parents and siblings, such that each person with the same family history is in the same configuration. The configuration is used as prior knowledge in calculating the posterior mean genetic liability to be able to get a more precise estimate of what the genetic liability should be for each person. The posterior mean genetic liability $E(\epsilon_{o,g}|z_o,z_{p1},z_{p2},\overrightarrow{z_s})$ uses the case/control status of the target $(z_o)$, the parents $(z_{p1}, z_{p2})$ and the siblings $(\overrightarrow{z_s})$. 
The case/control status of the family members informs us about what range the full liabilities lies in, since they have to be greater than $T$ if case and less than $T$ if control. The multivariate normal distribution above, doesn't include the targets own full liability, which will be affected by the targets own case/control status. We know from the liability threshold model assumptions that the targets full liability $\epsilon_{o,f} = \epsilon_{o,g}+\epsilon_{o,e}$. Including this gives the following distribution:
$$
\begin{pmatrix}
    \epsilon_{o,g} \\
    \epsilon_{o,f} \\
    \epsilon_{p1} \\
    \epsilon_{p2} \\
    \epsilon_{s}
\end{pmatrix}
\sim
MVN_5( \begin{pmatrix}
    0 \\
    0 \\
    0 \\
    0 \\
    0
\end{pmatrix}, \begin{pmatrix}
    h^2 & h^2 & 0.5h^2 & 0.5h^2 & 0.5h^2 \\
    h^2 & 1 & 0.5h^2 & 0.5h^2 & 0.5h^2 \\
    0.5h^2 & 0.5h^2 & 1 & 0  & 0.5h^2\\
    0.5h^2 & 0.5h^2 & 0 & 1  & 0.5h^2\\
    0.5h^2 & 0.5h^2 & 0.5h^2 & 0.5h^2 & 1
\end{pmatrix} )\\
$$
This is again with only one sibling, but can be expanded to an arbitrary number of siblings. Furthermore, based on this distribution the marginal distribution of a liability conditioned on the other liabilities can be found as shown in the example below.

Example:
For a family with no siblings we have: 

$$
\epsilon_{o,g}|\epsilon_{o,f},\epsilon_{p1},\epsilon_{p2} \sim N(\frac{(h^2-\frac{1}{2}h^4)\cdot \epsilon_{o,f} + (\frac{1}{2}h^2 - \frac{1}{2}h^4)\cdot (\epsilon_{p1}+\epsilon_{p2})}{1-\frac{1}{2}h^4}, \frac{h^2 - \frac{3}{2}h^4 + \frac{1}{2}h^6}{1-\frac{1}{2}h^4})
$$

where $\epsilon_{o,f}$ is the targets own full liability and $\epsilon_{p1}, \epsilon_{p2}$ is the parents full liability. By including the knowledge about the configuration, then we can sample the genetic liability from a normal distribution when the full liabilities have values that correspond to having the correct case/control status. This procedure is further explained in the Gibbs vignette.
