---
title: "Liability distributions with LT-FH"
output: rmarkdown::html_vignette
description: >
  An overview and explanation of the methodology used for drawing liabilities.
vignette: >
  %\VignetteIndexEntry{liability-distribution}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

This vignette shows how we model the liability distributions based on the work by Hujoel et al (2020).

## Simulation
In the simulations there are being simulated $n$ individuals with $m$ Single-Nucleotide Polymorphisms (SNPs), with $c$ causal SNPs. Each SNP is being drawn from a binomial distribution with ... and probability of the Minor Allele Frequency (MAF). Each SNP has its own MAF which is being drawn from a uniform distribution from $0.01$ to $0.49$. The effect size of the causal SNPs is drawn from a normal distribution with mean $0$ and variance $\frac{h^2}{c}$, and the effect size of non-causal SNPs is $0$.

## Liability threshold model
The liability threshold model assigns a phenotype to each person based on their full liability, which can be divided into two parts, namely genetic liability ($l_{g}$) and environment $l_{e}$. The genetic liability is calculated by multiplying the SNPs by their effect size and summing them $l_g = X_i \beta$, where $X_i$ is the i'th persons SNPs and $\beta$ is the effect sizes. It is assumed that $l_g \sim N(0,h^2)$, where $h^2$ is the heritability. The liability from the environment is sampled and is assumed that $l_e \sim N(0,1-h^2)$, which makes the full liability $l = l_g+l_e \sim N(0,1)$. The liability theshold model assigns a case to a person if $l\geq T$, with $P(l\geq T) = k$, where $k$ is the prevalence of trait, e.g $5%$ and T is the threshold of disease liability.

## LT-FH model
The LT-FH method relies on the liability threshold model, but expands it to be able to take case/control status and/or family history into consideration. 

