<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<!-- Inform modern browsers that this page supports both dark and light color schemes,
  and the page author prefers light. -->
<meta name="color-scheme" content="dark light">
<script>
  // If `prefers-color-scheme` is not supported, fall back to light mode.
  // i.e. In this case, inject the `light` CSS before the others, with
  // no media filter so that it will be downloaded with highest priority.
  if (window.matchMedia("(prefers-color-scheme: dark)").media === "not all") {
    document.documentElement.style.display = "none";
    document.head.insertAdjacentHTML(
      "beforeend",
      "<link id=\"css\" rel=\"stylesheet\" href=\"https://bootswatch.com/3/flatly/bootstrap.css\" onload=\"document.documentElement.style.display = ''\">"
    );
  }
</script>

<title>Function reference • geneference</title>

<!-- favicons -->
<link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png" />
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png" />
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png" />
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png" />

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- Flatly Theme - Light  -->
<link id="css-light" rel="stylesheet" href="https://bootswatch.com/3/flatly/bootstrap.css" media="(prefers-color-scheme: light), (prefers-color-scheme: no-preference)">

<!-- Darkly Theme - Dark -->
<link id="css-dark" rel="stylesheet" href="https://bootswatch.com/3/darkly/bootstrap.css" media="(prefers-color-scheme: dark)">

<!-- preferably CSS -->
<link rel="stylesheet" href="../preferably.css">
<link id="css-code-light" rel="stylesheet" href="../code-color-scheme-light.css" media="(prefers-color-scheme: light), (prefers-color-scheme: no-preference)">
<link id="css-code-dark" rel="stylesheet" href="../code-color-scheme-dark.css" media="(prefers-color-scheme: dark)">


<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Function reference" />
<meta property="og:image" content="/logo.png" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>




<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-index">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">geneference</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/geneference.html">Introduction</a>
</li>
<li>
  <a href="../reference/index.html">Functions</a>
</li>
<li>
  <a href="../articles/index.html">Vignettes</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
        <li>
  <a href="https://github.com/FireGutter/geneference/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
        
        


      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index">

    <colgroup>
      
      <col class="alias" />
      <col class="title" />
    </colgroup>

    <tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-simulation" class="hasAnchor"><a href="#section-simulation" class="anchor"></a>Simulation</h2>
          <p class="section-desc"><p>Simulate individuals and their genotypes, either including or excluding family history.</p></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="sim_fixed_family.html">sim_fixed_family()</a></code> </p>
        </td>
        <td><p>Simulation with fixed family size</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sim_no_family.html">sim_no_family()</a></code> </p>
        </td>
        <td><p>Simulation without family history</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sim_random_family.html">sim_random_family()</a></code> </p>
        </td>
        <td><p>Simulation with random family sizes</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sim_test.html">sim_test()</a></code> </p>
        </td>
        <td><p>Simulation directly in the R session without family history</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sim_varied_family.html">sim_varied_family()</a></code> </p>
        </td>
        <td><p>Simulation with varying family sizes</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-analysis" class="hasAnchor"><a href="#section-analysis" class="anchor"></a>Analysis</h2>
          <p class="section-desc"><p>Perform statistical analysis to discover what can be causally inferred from data. (tmp desc)</p></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="assign_GWAX_phenotype.html">assign_GWAX_phenotype()</a></code> </p>
        </td>
        <td><p>Create phenotype status by proxy</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="assign_ltfh_phenotype.html">assign_ltfh_phenotype()</a></code> </p>
        </td>
        <td><p>Calculate posterior mean genetic liabilities</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="analysis_association.html">analysis_association()</a></code> </p>
        </td>
        <td><p>Run association analysis</p></td>
      </tr>
    </tbody><tbody>
      
      <tr>
        <th colspan="2">
          <h3 id="section-import-calculate-tmp-subtitle-" class="hasAnchor"><a href="#section-import-calculate-tmp-subtitle-" class="anchor"></a>Import, calculate (tmp subtitle)</h3>
          <p class="section-desc"><p>Functionalities for importing data to the R session and performing a few calculations in R. (write better desc)</p></p>
        </th>
      </tr>
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="load_phenotypes.html">load_phenotypes()</a></code> </p>
        </td>
        <td><p>Import phenotype data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="load_results.html">load_results()</a></code> </p>
        </td>
        <td><p>Import and merge results</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="calculate_cov.html">calculate_cov()</a></code> </p>
        </td>
        <td><p>Calculate covariance matrix of liabilities</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="calculate_error_types.html">calculate_error_types()</a></code> </p>
        </td>
        <td><p>Calculate error types</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="augment_results.html">augment_results()</a></code> </p>
        </td>
        <td><p>Augment results for plotting</p></td>
      </tr>
    </tbody><tbody>
      
      <tr>
        <th colspan="2">
          <h3 id="section-visualisations" class="hasAnchor"><a href="#section-visualisations" class="anchor"></a>Visualisations</h3>
          <p class="section-desc"><p>Once you have imported data to R, you can use these functions for visualising stuff. The functions listed here can either return a ggplot object or save an image of the plot to disk. Users familiar with the ggplot2 package can manipulate the returned objects. (tmp desc)</p></p>
        </th>
      </tr>
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="plot_estimates_vs_true.html">plot_estimates_vs_true()</a></code> </p>
        </td>
        <td><p>Plot true effects against estimated effects of SNPs</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plot_manhattan.html">plot_manhattan()</a></code> </p>
        </td>
        <td><p>Manhattan plot of p-values</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plot_pmgl_vs_true.html">plot_pmgl_vs_true()</a></code> </p>
        </td>
        <td><p>Plot true genetic liabilities against posterior mean genetic
liabilities</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plot_pval_QQ.html">plot_pval_QQ()</a></code> </p>
        </td>
        <td><p>Plot a QQ-plot of p-values</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plot_pval_hist.html">plot_pval_hist()</a></code> </p>
        </td>
        <td><p>Plot a histogram of p-values</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="compare_beta.html">compare_beta()</a></code> </p>
        </td>
        <td><p>Plot transformed linear GWAS beta values against the LTFH_transformed
values.</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-helper-functions" class="hasAnchor"><a href="#section-helper-functions" class="anchor"></a>Helper functions</h2>
          <p class="section-desc"><p>These are miscellaneous helper functions also exported by the package. Write one line motivating their use.</p></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="rmnorm.html">rmnorm()</a></code> </p>
        </td>
        <td><p>Multivariate normal distributions</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="convert_geno_file.html">convert_geno_file()</a></code> </p>
        </td>
        <td><p>Convert files to .bed</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="covmatrix.html">covmatrix()</a></code> </p>
        </td>
        <td><p>Create covariance matrix for liabilities</p></td>
      </tr>
    </tbody>
    </table>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by FireGutter, Jacob Birkeland Steenstrup, Rasmus Brostrøm Jensen, Magnus Conradsen, Marc Timmer Jensen.</p>
</div>

<div class="pkgdown">
  <p>Made with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1, using <a href="https://preferably.amirmasoudabdol.name/?source=footer">preferably</a> template.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


